azureevaluator.py

```diff
@@ -1,27 +1,201 @@
+
+		return judge_result
 
 class AzureVMSS:
 	"""
-	An Azure Virtual Machine Scale Set.
+	An Azure Virtual Machine Scale Set. A Set contains a single machine type.
 	"""
+	machine_type: 'MachineType'
+	azurevmss_name: str
+	azurevm_dict: Dict[str, 'AzureVM']
+	vmss: VirtualMachineScaleSet
+	azure: Azure
+	
+	def __init__(self, machine_type: MachineType, azurevmss_name: str, vmss: VirtualMachineScaleSet , azure: Azure):
+		self.machine_type: machine_type
+		self.azurevmss_name: azurevmss_name
+		self.azurevm_dict = {}
+		self.vmss = vmss
+		self.azure = azure
+
 	async def submit(self, judge_request: JudgeRequest) -> JudgeResult:
-		# TODO: get the right VM, or scale up if not available.
-		pass
+		resource_allocation = judge_request.resource_allocation
+
+		# Get a right vm that is available
+		vm = self.check_available_vm(resource_allocation)
+
+		# If no available vm than add capacity
+		if vm is None:
+			# Get available vm after the added capacity, error if no available
+			self.add_capacity()
+
+			vm = self.check_available_vm(resource_allocation)
+
+			if vm is None:
+				# Throw exception
+				raise Exception("No vm available for judge request, even after adding capacity")
+
+		# Submit using the vm the judge request
+		judge_result = await self.submit_vm(vm, judge_request)
+
+		# Reduce capacity and update vm dict
+		self.reduce_capacity()
+		self.update_vm_dict()
+
+		return judge_result
+
+	async def add_capacity(self):
+		# Increase capacity of vmss with an arbitrary max of 5
+		capacity = self.vmss.sku.capacity
+		if capacity < 5:
+			await self.azure.set_capacity(capacity + 1, self.azurevmss_name)
+		
+		# Update vm_dict, vm(s) could have been added
+		self.update_vm_dict()
+			
+	async def reduce_capacity(self):
+		# Decrease capacity of vmss
+		capacity = self.vmss.sku.capacity
+
+		# There needs to be capacity to be removed
+		if capacity > 0:
+			await self.azure.set_capacity(capacity - 1, self.azurevmss_name)
+
+		# Update vm_dict, vm(s) could have been deleted
+		self.update_vm_dict()
+	
+	async def submit_vm(self, vm: VirtualMachineScaleSetVM, judge_request: JudgeRequest) -> JudgeResult:
+		azurevm = self.vm_dict[vm.name]
+		judge_result = await azurevm.submit(self, judge_request)
+
+		return judge_result
+	
+	async def check_available_vm(self, resource_allocation: ResourceSpecification) -> VirtualMachineScaleSetVM:
+		# Get the list of vms
+		vms = self.azure.list_vms(self.machine_type, self.azurevmss_name)
+
+		# Update vm_dict, make sure the dict is up to date
+		self.update_vm_dict()
+
+		# Go over the virtual machine to find one with enough capacity
+		for vm in vms:
+			# Get the azure vm class instance associated to the vm
+			if self.vm_dict[vm.name]:
+				azurevm = self.vm_dict[vm.name]
+			
+			# Check if there is enough free resource capacity on this vm
+			if azurevm.capacity(resource_allocation):
+				
+				return vm
+		
+		# No vm found
+		return None
+	
+	async def update_vm_dict(self):
```

Comment:
This function is internal, right? (not meant to be used outside the class itself)

Cause then typical Python conventions is to prefix it with one or two `_`